### DISCUSSION TITLE
Permit usage of the `default` literal as a pattern

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#5704] [Language Ideas]
After an extensive research, I could not find a single discussion or issue about this topic, so I decided to open a new one. It could be a duplicate, in which case it will be marked as answered.

Currently, the compiler emits error CS8505, prohibiting usage of the `default` literal in an `is` expression. However, `default(T)` is permitted. The `default` can easily have its type inferred from the type of the expression that is being evaluated in the matching.

This can be a tad annoying in quite a few scenarios, including:
- generic type parameters, not constrained to a class or struct (`x is null` will not compare against `default` if `x` is a struct)
- structs
  - non-primitive structs
  - `char`, my personal favorite, whose default value literal is `'\0'`, and using `c is 0` will not work if `c` is a `char`

**Comment by jnm2** [comment]
> `char`, my personal favorite, whose default value literal is `'\0'`, and using `c is 0` will not work if `c` is a `char`

I vastly prefer `someChar is '\0'` over `someChar is default` or `someChar is 0`. For the same reason that I prefer `is null`, `is false`, and `is 0m` etc over `is default`.

`is default` would be nice for non-primitive structs though for sure.
